### ISSUE TITLE
x/build/cmd/release: evaluate builders used for upcoming major Go release

### ISSUE THREAD

**Issue by dmitshur** [golang/go#40561]
At the start of each code freeze, we should evaluate the builders used to make the upcoming major Go release, and confirm they are the most appropriate choice out of what is available.

The start of a code freeze is a safe time to introduce a builder change, as it will be tested during pre-release versions, starting with beta 1 and onwards.

Making this a release blocker for Go 1.16 (and not okay after beta 1—it needs to be done before). Once completed, this issue should be moved to the next major release milestone.

/cc @golang/osp-team

**Comment by dmitshur** [comment]
The @golang/release team has considered the current builders, and we are proposing the following builder changes for the Go 1.16 release:

go1.16*.linux-amd64.tar.gz
--------------------------

|         | Builder Name       | Host Type          |
|---------|:-------------------|:-------------------|
| Current | linux-amd64-jessie | host-linux-jessie  |
| **New** | linux-amd64        | host-linux-stretch |

(Or rather a new builder, identical to the `linux-arm` one, just named `linux-amd64-stretch` instead.)

**Rationale:** Debian 8 "Jessie" Long Term Support (LTS) has ended on June 30, 2020 (see https://wiki.debian.org/LTS). We want to start using the next stable release, Debian 9 "Stretch", which has LTS support until June 2022.

Note that this has caused #31293 in April 2019 when this change was made unintentionally as part of a minor release, and it was reverted. We hope that it should be okay to try again for Go 1.16 a year later, since this is being done intentionally this time, and for a major release. CC @ianlancetaylor, @jayconrod, @bradfitz.

go1.16*.linux-armv6l.tar.gz
---------------------------

|         | Builder Name  | Host Type               |
|---------|:--------------|:------------------------|
| Current | linux-arm     | host-linux-arm-scaleway |
| **New** | linux-arm-aws | host-linux-arm-aws      |

**Rationale:** The replacement builder has proven to be more reliable, and it is easier for us to maintain.

go1.16*.linux-arm64.tar.gz
--------------------------

|         | Builder Name       | Host Type               |
|---------|:-------------------|:------------------------|
| Current | linux-arm64-packet | host-linux-arm64-packet |
| **New** | linux-arm64-aws    | host-linux-arm64-aws    |

**Rationale:** The replacement builder has proven to be more reliable, and it is easier for us to maintain.

---

Feedback is welcome: we are happy to alter the plan based on information we haven't already considered. These builder changes are planned for Go 1.16 Beta 1, and we will re-evaluate this as needed based on any new findings and feedback.

**Comment by dmitshur** [comment]
[CL 276034](https://golang.org/cl/276034) has implemented the aforementioned plan for Go 1.16.

We'll need to revisit this issue next time during the Go 1.17 development cycle. (If we discover problems with the Go 1.16 plan, we'll come back here, or file new issues.) Moving to the next milestone for now.

**Comment by thanm** [comment]
While working on issue #39326, it has become clear to me that the C compile we have installed on our windows gomotes (at least the ones I have used) is pretty ancient -- GCC 5.   It would be great if we could update the version to something a bit more recent (V10 maybe).  Also worth noting that many folks are now using clang instead of gcc on windows -- it would be great if our gomotes had clang available as well.

**Comment by dmitshur** [comment]
@thanm Can you please file a separate issue to track that work, and CC @golang/release? It may need more discussion/investigation. Thanks.

**Comment by thanm** [comment]
Filed https://github.com/golang/go/issues/43195. Thanks.

**Comment by cagedmantis** [comment]
The @golang/release team has considered the current builders, and we are proposing the following builder change for the Go 1.17 release:

go1.17.*.freebsd-amd64.tar.gz
--------------------------

|         | Builder Name       | Host Type          |
|---------|:-------------------|:-------------------|
| Current | freebsd-amd64-11_2   | host-freebsd-11_2 |
| **New** | freebsd-amd64-11_4  | host-freebsd-11_4 |

**Rationale:** Using the latest "patch" release of the oldest supported "major" release of FreeBSD.

https://github.com/golang/go/issues/45727#issuecomment-825803454

go1.17.*.freebsd-386.tar.gz
--------------------------

|         | Builder Name       | Host Type          |
|---------|:--------------------|:-------------------|
| Current | freebsd-386-11_2    | host-freebsd-11_2  |
| **New** | freebsd-386-11_4   | host-freebsd-11_4 |

**Rationale:** Using the latest "patch" release of the oldest supported "major" release of FreeBSD.

https://github.com/golang/go/issues/45727#issuecomment-825803454

**Comment by dmitshur** [comment]
There haven't been noteworthy builder changes in the last 2 weeks since https://github.com/golang/go/issues/40561#issuecomment-830424055. I've looked over our current selection of releaselets for 1.17 once again, and I think it's good to proceed with.

I did spot one opportunity we can consider: to start using `darwin-amd64-11_0` (Big Sur) builder for 1.17, in line with our general approach of using latest macOS builders for macOS releases. (And we have the opportunity of testing it during Beta 1).

Let's make a decision on this one entry. Otherwise, I think it's okay to look again for the 1.18 cycle.

**Comment by dmitshur** [comment]
We agreed to try `darwin-amd64-11_0` for 1.17 onwards, filed #46161 for that, and moving this to 1.18 milestone.

**Comment by cagedmantis** [comment]
The @golang/release team has considered the current builders, and we are proposing the following builder change for the Go 1.18 release:

|  Port  |  Current Builder  | Current Host Type  | Proposed Builder  | Proposed Host Type |
|----------|-----------------|---------------------|------------------|----------------------|
| freebsd-386  |  freebsd-386-11_4  | host-freebsd-11_4 |  freebsd-386-12_2 | host-freebsd-12_2 |
| freebsd-amd64  |  freebsd-amd64-11_4  | host-freebsd-11_4 |  freebsd-amd64-12_2 | host-freebsd-12_2 |
| windows-386     | windows-386-2008 | host-windows-amd64-2008 | windows-386-2012 | host-windows-amd64-2012 |
| windows-amd64 | windows-amd64-2008 | host-windows-amd64-2008 | windows-amd64-2012 | host-windows-amd64-2012 |

**Rationale:** Using the latest "patch" release of the oldest supported "major" release of FreeBSD. Using the latest supported version of Windows server.

**Comment by cagedmantis** [comment]
The @golang/release team has revised the proposed releaselets to be used for the Go 1.18 release:

|  Port  |  Current Builder  | Current Host Type  | Proposed Builder  | Proposed Host Type |
|----------|-----------------|---------------------|------------------|----------------------|
| freebsd-386  |  freebsd-386-11_4  | host-freebsd-11_4 |  freebsd-386-12_2 | host-freebsd-12_2 |
| freebsd-amd64  |  freebsd-amd64-11_4  | host-freebsd-11_4 |  freebsd-amd64-12_2 | host-freebsd-12_2 |

**Rationale:** Using the latest "patch" release of the oldest supported "major" release of FreeBSD.  The Windows change was removed because of the discussions in #47845 and release team meetings.

**Comment by dmitshur** [comment]
By now we have darwin builders with macOS 12 both on amd64 and arm64 architectures (#49149), though that work is still evolving. We will consider whether to use them or still use macOS 11 for building 1.18 before Beta 1. The current trajectory is to use macOS 12 builder at least for arm64, but that may change. (Feedback on this is welcome.)

**Comment by dmitshur** [comment]
We've discussed macOS today and agreed to use macOS 12 releaselets both for amd64 and arm64 architectures starting with Go 1.18 Beta 1. Send [CL 369958](https://go.dev/cl/369958).

**Comment by dmitshur** [comment]
All looks ready for Go 1.18 Beta 1, so moving this recurring issue to the next milestone.

**Comment by heschi** [comment]
I've updated 1.18 to use FreeBSD 12.3, which has the fix for the memory corruption bug discussed in #49209.

**Comment by heschi** [comment]
Current state of affairs is visible at https://cs.opensource.google/go/x/build/+/master:internal/releasetargets/releases.txt;l=105;drc=14add1f2a409547572304844b265b007589d3c16.

Linux: We've updated most of the linux builders to Bullseye, but I suspect not arm-aws, and I'm not sure of the state of arm on GCE (@cagedmantis).

macOS: We've updated AMD64 to 13, but don't have an ARM64 13 builder yet. That's slightly unfortunate but I don't think it's a major problem.

Windows: seems like we want to be on 2008 for as long as that's our minimum supported version?

FreeBSD: should probably be updated to 13.

Anything I'm missing?

**Comment by cagedmantis** [comment]
`linux-arm-aws` is currently on Buster. We do not have immediate plans to deploy a replacement on GCE nor do we plan to update the current image.

**Comment by heschi** [comment]
Nothing else to do this cycle, bumping.

**Comment by heschi** [comment]
Linux: bullseye is still the latest version. linux-arm-aws is still behind and we're still okay with it.
macOS: 13 is still the latest version, so we're good on AMD64 and still behind on ARM64. I think we're still okay with it.
Windows: we've dropped support for Windows 8 -- we need to target windows-386/amd64-2016. Windows 11 is still the most recent major version on ARM64, though it's an old patch level by now.
FreeBSD: should perhaps be updated to 13.2 but it's a second-class port and not our problem.

**Comment by heschi** [comment]
Done for this cycle.

**Comment by dmitshur** [comment]
Done for the Go 1.23 cycle.

**Comment by dmitshur** [comment]
[CL 612555](https://go.dev/cl/612555) regenerated the list for Go 1.24. It's the same list as for Go 1.23 but without windows/arm (32-bit), and that's noted at https://go.dev/doc/go1.24#windows.

I checked again now that we're post-freeze, before RC 1, and there haven't been further changes (CC @mknyszek). Unless some new issue is uncovered during pre-release testing, this is done for Go 1.24 cycle; moving to next.

**Comment by dmitshur** [comment]
[CL 664215](https://go.dev/cl/664215) generated a preliminary list for Go 1.25. It's the same list as for Go 1.24 except the macOS minimum version became 12, and that's noted at https://tip.golang.org/doc/go1.25#darwin.

I checked again now that that it's the freeze day, and there haven't been further changes. Unless some new issue is uncovered during pre-release testing, this is done for Go 1.25 cycle.

**Comment by cherrymui** [comment]
This list still looks good to me for Go 1.26. If we uncover anything during pre-release testing, we can revisit. Moving to Go 1.27 for now.

**Comment by dmitshur** [comment]
[CL 767441](https://go.dev/cl/767441) generated a preliminary list for Go 1.27. It's the same list as for Go 1.26 except the macOS minimum version became 13, and that's noted at https://tip.golang.org/doc/go1.27#darwin.

I re-checked now that we're in the release freeze, and there haven't been changes since then. If we uncover something during pre-release testing, we can revisit, otherwise moving this to the next milestone.
